### PR TITLE
App definition for Inventory API

### PR DESCRIPTION
**Pull request by ian-schu**
Jira: https://auctane.atlassian.net/browse/PLAT-2838

## Why / What?
An initial definition for Inventory Apps on the Connect platform. This PR is only the OpenAPI yaml spec.

# At a glance
Current as of evening of 6/2/21 ...

1. **Metadata**\
Inventory Apps will need to specify what parts of this contract they can fulfill. Not every Inventory App will support a `fetch-delta`, for instance – that's mostly only Shopify. These capabilities will be part of the metadata registry, and _are not shown in the yaml file in this PR_.
2. **Fetching stock levels**\
Clients can hit any of the following to pull inventory from the 3rd party source:
    - `/fetch_inventory_full` for pulling in everything that an inventory source has available
    - `/fetch_inventory_partial` for getting only inventory levels for specific SKUs / identifiers
    - `/fetch_inventory_delta` for getting only inventory levels that have changed since a given date
3. **Pushing stock levels**\
Basically just:
    - `/push_inventory`
4. **Long-running / async operations**\
Depending on the inventory source, fetching and/or pushing can be long-running operations. In these cases, Inventory Apps will send back a `pending` response, along with an opaque `cursor` and a suggested `poll_after_seconds`. Clients should include the `cursor` in their next request, and hit the same endpoint after the suggested delay. Inventory Apps will use the `cursor` as a locator for the operation that's already in-flight, so there isn't a need to resend _data_ in these subsequent polling requests.

I went through several variations to arrive at this current pattern of handling long-running async ops. Motivations / reasoning will be easier to express in more real-time conversational form.

### CODE REVIEW

**Comment by ian-schu** [comment]
## Notes for review / later iteration:
1. **Handling asynchronous / long-running tasks**
We know from experience in both `consolidated-inventory` and at ShippingEasy that some store platforms respond _asynchronously_ to push / pull requests.
\
In these cases we're usually left polling the store platform to see when the data is ready, or else providing a notification URL to receive a callback when processing is done. (The latter approach is supported in some form by Amazon and Shopify at least, but to my knowledge has not been implemented anywhere yet).
\
This scenario is handled only dimly / implicitly in the current YAML specification. It's _plausible_ that the arbitrary `context` object could return a `polling_url` or something. But maybe it would be better to imagine that Inventory Apps can _take_ a webhook URL from consumers, and simply hit the webhook when fetching is done? Essentially, expecting that Inventory Apps will not obligate consumers to do their own polling, but will encapsulate that behavior and make life easy for consumers.

2. **Context / pagination, etc.**
Coming from `consolidated-inventory` where we called directly to various store / inventory integrations, it was fine to lump arbitrary / transient request metadata into the `context` blob. This included things like `per_page`, `total_pages`, `nextCursor`, and so on.
\
In the world to come where the consumers of this API are one more step removed, I'm not sure that makes sense anymore. Will need some combination of time + feedback in order to get out of the consolidated-inventory mindset and think more like a consumer of this API. What will consumers want to see / provide regarding pagination, cursors etc?

File: packages/inventory-api/spec.yaml [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+          type: integer
+          example: 1
+
+    InventoryFetchItem:
+      description: The reported stock levels from an inventory source when fetching
+      allOf:
+        - $ref: '#/components/schemas/InventoryItemBase'
+        - type: object
+          properties:
+            locationId:
```
Comment (marcusmateus) [inline packages/inventory-api/spec.yaml:143]: This should be `warehouseId`. Outside of Easy the standard is that "location" refers to a bin in a warehouse.

Comment (ian-schu) [reply packages/inventory-api/spec.yaml:143]: Oh yeah, thank you.

File: packages/inventory-api/spec.yaml [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+      required: true
+      content:
+        application/json:
+          schema:
+            type: object
+            properties:
+              context:
+                $ref: '#/components/schemas/InventoryFetchContext'
+              credentials:
+                $ref: '#/components/schemas/RequestCredentials'
```
Comment (marcusmateus) [inline packages/inventory-api/spec.yaml:213]: It's been a while, but I seem to remember the Connect team saying there was some standard way of handling credentials. Probably worth double-checking.

File: packages/inventory-api/spec.yaml [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+components:
+  schemas:
+    InventoryItemBase:
+      type: object
+      description: Common / shared properties for InventoryItems, regardless of whether receiving data (fetch) or sending it (push) 
+      required:
+        - sku
+        - integrationEntityId
+        - updatedAt
```
Comment (marcusmateus) [inline packages/inventory-api/spec.yaml:108]: If I recall, not all providers include an `updatedAt` concept so required might not work... wonder if perhaps we should have 2 timestamps... the other being `fetchedAt` or similar that is required, for when the retrieval happened.

File: packages/inventory-api/spec.yaml [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        - type: object
+          properties:
+            sku:
+              type: string
+              example: t-shirt-small-red
+              description: The merchant-supplied identifier for the product.
+            integrationEntityId:
+              type: string
+              example: t-shirt-small-red
+              description: The unique inventory record in the *source* system. This will often be the same as the SKU, but could be different in inventory source platforms that allow duplicate SKUs.
```
Comment (marcusmateus) [inline packages/inventory-api/spec.yaml:160]: Both `sku` and `integrationEntityId` seem to be included in `InventoryItemBase`. Any reason for re-including them? Not even the descriptions changed.

File: packages/inventory-api/spec.yaml [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    InventoryFetchContext:
+      # This is relevant for the integration code inside the app, but I'm unclear how relevant it is for consumers facing this API.
+      description: An open-ended JSON blob which includes (usually) transient pagination details relative to the inventory source.
+      type: object
+      additionalProperties: true
+      example: { page: 2, per_page: 500 }
+
+    RequestCredentials:
+      type: object
+      description: Credentials necessary to send authenticated requests to the inventory source – these will vary depending on the type of the inventory provider.
```
Comment (marcusmateus) [inline packages/inventory-api/spec.yaml:171]: Think you need `additionalProperties: true` here too.

File: packages/inventory-api/spec.yaml [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+          type: string
+          example: Throttled due to exceeding available query capacity
+        details:
+          description: Additional detail about what went wrong.
+          type: string
+          example: Calculated cost was 5400, but only 1200 units are currently available. (Restore Rate 100)
+        retry_after_seconds:
+          description: A suggested wait interval before the request should be retried, if applicable.
+          type: number
+          example: 42
```
Comment (marcusmateus) [inline packages/inventory-api/spec.yaml:196]: I like this error standardization.

File: packages/inventory-api/spec.yaml [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+              inventoryItems:
+                type: array
+                items:
+                  $ref: '#/components/schemas/InventoryFetchItem'
+              context:
+                $ref: '#/components/schemas/InventoryFetchContext'
+              hasMoreData:
+                type: boolean
+
+    # This is mostly borrowed from Freight API, but a good starting point
```
Comment (marcusmateus) [inline packages/inventory-api/spec.yaml:280]: Not familiar, where is this?

File: packages/inventory-api/spec.yaml [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    FetchInventoryRequestBase:
+      description: Request body for making a basic (full) inventory request
+      required: true
+      content:
+        application/json:
+          schema:
+            type: object
+            properties:
+              context:
+                $ref: '#/components/schemas/InventoryFetchContext'
```
Comment (marcusmateus) [inline packages/inventory-api/spec.yaml:211]: Perhaps should have some explanation / documentation somewhere that "context" should be handed back on subsequent calls until `hasMoreData` is false.

File: packages/inventory-api/spec.yaml [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+          schema:
+            properties:
+              inventoryItems:
+                type: array
+                items:
+                  $ref: '#/components/schemas/InventoryFetchItem'
+              context:
+                $ref: '#/components/schemas/InventoryFetchContext'
+              hasMoreData:
+                type: boolean
```
Comment (marcusmateus) [inline packages/inventory-api/spec.yaml:278]: In Consolidated Inventory we hacked in an `isPolling` function to differentiate between when more data was immediately available and when a delay was needed... I think we should be more explicit here.
https://github.com/ShippingEasy/consolidated-inventory/blob/bad87709a0714e9ad9861064c548fb1282a93ab2/apps/bg-job/src/inventory/gateways/types.ts#L27-L29

File: packages/inventory-api/spec.yaml (hunk 2 above)
Comment (jeversmann-shipstation) [reply packages/inventory-api/spec.yaml:213]: I think it would be good if we shared as much as possible with https://github.com/ShipEngine/connect/blob/master/packages/order-source-api/spec.yaml#L927-L950
But it's not necessarily a requirement to re-use the entire `auth` object model. I'd prefer if we put things that are "the same" in the same fields 🤷🏽

File: packages/inventory-api/spec.yaml [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+      operationId: PushInventory
+      tags:
+        - Push Inventory
+      description: Send new or updated stock quantities to the inventory provider or store platform
+      requestBody:
+        $ref: '#/components/requestBodies/PushInventoryRequestBody'
+      responses:
+        '200':
+          description: Success
+          # What content do we care about receiving back here?
```
[the remark this answers is not in the document]
Comment (jeversmann-shipstation) [reply packages/inventory-api/spec.yaml:70]: I also think it would be good to include an array by `sku` here in order to be able to capture partial success/failure responses for integrations which can't perform bulk updates on the 3rd party.

**Review by jeversmann-shipstation — COMMENTED** [review]
## Handling asynchronous / long-running tasks
I think that providing an arbitrary url for polling transaction results puts a burden on the client of this API to understand interacting with the outside world. I would like to follow a structure similar to the "Delayed Notifications" [proposal](https://github.com/shipstation/integrations-ecommerce/pull/433) for OrderSourceAPI where `result` is an enum instead of a boolean, and on `result=pending` the client should know to call a `/status` endpoint with an ID supplied in the response.

For 3rd parties who surface this data to us via a polling endpoint, the module could pass the request back through to the 3rd party equivalent endpoint and translate the response to a canonical "result" status. For webhook/SQS based 3rd party APIs we'll need to build more infrastructure level support in Connect to handle, but we've got some experiments ongoing with Amazon that we can talk about if you're interested.

## Context / pagination, etc.
I put my primary thoughts on this in line, but I think a single, opaque `cursor` field is the way go to. The old connect sdk based order source runtime does this by [JSON serializing a bunch of stuff](https://github.com/ShipEngine/connect/blob/master/packages/order-source-runtime/src/mapping/sales-orders-export/index.ts#L8-L22) but ultimately the serialization method & payload contents is an implementation detail of the integration.

**Comment by jeversmann-shipstation** [comment]
Also the other specs are all snake cased! Please don't do this to me 😭

File: packages/inventory-api/spec.yaml (hunk 7 above)
Comment (jeversmann-shipstation) [reply packages/inventory-api/spec.yaml:280]: https://github.com/ShipEngine/connect/blob/master/packages/freight-api/spec.yaml

File: packages/inventory-api/spec.yaml [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+                type: object
+                properties:
+                  supported_fetch_types:
+                    type: array
+                    example: [full, delta]
+                    items:
+                      type: string
+                      enum: [full, partial, delta]
+                  supports_inventory_push:
+                    type: boolean
```
[the remark this answers is not in the document]
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:97]: Oh very nice. Still have not circled back around to read about what `metadata` actually looks like in this world, but you're starting to answer that question implicitly here.

My conception of this is: for a given implementation of this API (e.g. for a given Inventory App) the data returned from this endpoint would be constant, never-changing.

So an eventual app implementation for Shopify, for example, would _always_ return something like:
```javascript
{
  supported_fetch_types: ['full', 'partial', 'delta'],
  supports_inventory_push: true
}
```

... or so I think 🤔 

I suppose that makes this a good fit for static metadata, yeah?

File: packages/inventory-api/spec.yaml (hunk 4 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:160]: Oh, probably leftover from before I had the shared base schema. I'll fix this 👍🏻 thanks.

**Comment by marcusmateus** [comment]
> I would like to follow a structure similar to the "Delayed Notifications" proposal for OrderSourceAPI where result is an enum instead of a boolean, and on result=pending the client should know to call a /status endpoint with an ID supplied in the response.

I like this strategy... and it's great that y'all are already considering ways for registering for updates via webhooks, etc. rather than polling.

> opaque cursor field is the way go to

I agree whether it is "cursor" or "context" or whatever name is standard... we should look at it as an opaque bag of data.

File: packages/inventory-api/spec.yaml (hunk 5 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:171]: Nah, turns out there a many ways to Rome here 🙂 

image

File: packages/inventory-api/spec.yaml (hunk 6 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:196]: A lot of it was inspired by the error schema that I found nearby in `/freight-api` 😏 
Credit where it's due.

File: packages/inventory-api/spec.yaml (hunk 2 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:213]: I was not aware, but I will be happy to standardize here 👍🏻

File: packages/inventory-api/spec.yaml (hunk 8 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:211]: To my mind, this is still very much "unfinished business" per my NOTES comment at the top-level of the PR. I agree with you though – if we do end up keeping this, we should document what it's there for and how consumers should use it.

For me I just want to make sure I'm not cognitively anchored to consolidated-inventory too much, so I'm holding this very lightly. In other words I have no confidence that these two things should bear a strong resemblance to one another:
1. the way `consolidated-inventory` handles a `context` coming back from one of its inventory-gateways
2. the way consumers of this API 👇🏻 ☝🏻 should handle a `context` blob that may or may not come back from an inventory provider ... which is supposed to be abstracted by the inventory app.

File: packages/inventory-api/spec.yaml (hunk 7 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:280]: More relevant to your question might _why_ I borrowed from the Freight API, and the answer to that is: they are serving as the first implementation of some new app-type that goes outside of the established molds of Carrier or Order Source apps. Or at least, they're the latest contribution that Josh directed me to treat as prior art. Which has been helpful 👍🏻

File: packages/inventory-api/spec.yaml (hunk 9 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:278]: ... Makes me wonder if we want to follow the pattern of `retry_after_seconds` but in this case `retry` would be more like `poll`.

... But again, this goes to my top-level comment about asynchronous fetch handling. Should we be obligating our consumers to poll, or just hide that behavior inside of inventory apps?

Another area where I think I'm still getting my head around the two-levels-deep abstraction. Things might be clearer next week after some simmering.

File: packages/inventory-api/spec.yaml [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+              example: t-shirt-small-red
+              description: The merchant-supplied identifier for the product.
+            integrationEntityId:
+              type: string
+              example: t-shirt-small-red
+              description: The unique inventory record in the *source* system. This will often be the same as the SKU, but could be different in inventory source platforms that allow duplicate SKUs.
+
+    InventoryFetchContext:
+      # This is relevant for the integration code inside the app, but I'm unclear how relevant it is for consumers facing this API.
+      description: An open-ended JSON blob which includes (usually) transient pagination details relative to the inventory source.
```
[the remark this answers is not in the document]
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:164]: Point taken, and per another comment from Marcus I think we're on the same page that there's more work to be done on `context` and pagination. The term 'context' is purposefully vague because within `consolidated-inventory` this blob took on multiple concerns from the individual gateways / integrations that we called to. I am not nearly convinced that we need to keep anything nearly so vague at this level of abstraction. We might be able to replace this with a `pagination` key or a `nextCursor` key or something.

File: packages/inventory-api/spec.yaml (hunk 3 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:108]: Good call, I'll do some more work here.

File: packages/inventory-api/spec.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:70]: :nodding: good thoughts, I'll get something in here as a first attempt.

**Comment by ian-schu** [comment]
> Also the other specs are all snake cased! Please don't do this to me 😭

@jeversmann-shipstation I'm happy to fall in line with convention 🙂 don't assume any aspect of code-style here was done on purpose.

So just stealing a look at some other specs ... you mean all `properties` are snake-cased, right?
I can deal with making that switch, no problem.

**Comment by jeversmann-shipstation** [comment]
Yea, JSON properties and endpoint paths are snake cased, things like the model names in the spec can be whatever. 👍🏽

**Comment by ian-schu** [comment]
@jeversmann-shipstation thanks for the good thoughts here. I realized I waded into your inline commentary before seeing the top-level view.

https://github.com/ShipEngine/connect/pull/140#pullrequestreview-671634805

Quickly though:
1. I like the opaque `cursor` 👍🏻 
2. I _love_ the more-deterministic, enum-based return shape for various states of fetching.

File: packages/inventory-api/spec.yaml (hunk 9 above)
Comment (marcusmateus) [reply packages/inventory-api/spec.yaml:278]: Pretty sure going with Josh's `result` enum and `/status` endpoint thoughts below serve to address the need in my comment.

File: packages/inventory-api/spec.yaml (hunk 5 above)
Comment (marcusmateus) [reply packages/inventory-api/spec.yaml:171]: Then, for consistency, perhaps remove it above?

File: packages/inventory-api/spec.yaml (hunk 2 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:213]: Having looked at this more closely ... Yes I'll be happy to use `auth` for the name of the key here. But other than the key, very little else here can be standardized, because there's essentially zero reliable overlap in the properties that end up in this blob.

Here's a sampling of several of the `XYZCredentials` types that we use for the various inventory providers. Note the lack of regularity.

image

So, I don't think we have a ton of room for improvement here.

File: packages/inventory-api/spec.yaml (hunk 5 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml]: Good catch, done 👍🏻

File: packages/inventory-api/spec.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:97]: On closer inspection, I see what I think is a recent example from Freight API, here:
https://github.com/ShipEngine/connect/blob/master/packages/freight-api/src/app/freight-app-metadata.ts

... And after some layers of indirection, it looks like values of this type ultimately get fed into `ConnectRuntimeApp.data`.

Is that the size of what you're suggesting, or is there more to it?

To be clear:
- ✅  I'm pretty sure that all capabilities will be consistent across connections for the same integration. As you said.
- ❓ But I'm basically asking what else is involved with defining metadata for the integrations registry – if more than what I described above.

[the remark this answers is not in the document]
Comment (jeversmann-shipstation) [reply packages/inventory-api/spec.yaml:97]: The official "source of truth" for the shape of an integration's registry data lives in [integrations-platform](https://github.com/shipstation/integrations-platform/tree/master/Core) instead of inside of Connect, but since I don't think there's been movement on what the shape of inventory integrations should be in the registry, you could use this draft to set some assumptions for what it will eventually contain that we can base the design off of.

[the remark this answers is not in the document]
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:97]: Got ya.

And so ... is it mostly a matter of modeling _what varies_ between different implementations of this common Inventory API? In other words, what I've already sketched out – supported types of fetching and pushing ... do you suppose this is a reasonable starting point for modeling inventory integrations in the registry? Or does the registry want to know more than that?

[the remark this answers is not in the document]
Comment (jeversmann-shipstation) [reply packages/inventory-api/spec.yaml]: Yea, I think the data you've described all seems like good candidates for the registry, along with anything else that the GTM/platform teams need in order to make their behavior "data driven" so they won't need to write any code that's specific to a single integration.

[the remark this answers is not in the document]
Comment (ian-schu) [reply packages/inventory-api/spec.yaml]: 👍🏻 
Thanks.

**Comment by ian-schu** [comment]
OK @jeversmann-shipstation following up on this now that I'm starting to implement it, at least in the spec ...

> I would like to follow a structure similar to the "Delayed Notifications" proposal for OrderSourceAPI where result is an enum instead of a boolean, and on result=pending the client should know to call a /status endpoint with an ID supplied in the response.

... Question is, what happens when a client hits `/status` and finds out that their pending request is ready to be downloaded? 
1. Do they actually get the data they originally requested, but via hitting `/status`? Because, that feels weird / wrong.
2. Or in our case, would the client be obliged to hit one of our `fetch_inventory_whatever` endpoints again? That also feels weird, because those endpoints – as they're currently envisioned – are meant for kicking off _new_ fetch requests, not getting the results of existing ones.
3. I could imagine _maybe_ the latter scenario isn't so weird if clients are given an opaque `cursor` which actually deserializes to something the inventory app understands as "download the results of this particular long-running fetch operation". But I guess I had imagined that a `cursor` would mainly be for _paging_ through a series of responses that are immediately available; in this case it would be for accessing the results of an asynchronous fetch that are now hopefully available. Maybe I shouldn't sweat the flexibility / reuse here? Tell me if this does / doesn't trigger your spidey sense.

File: packages/inventory-api/spec.yaml [hunk 13]
```diff
@@ -0,0 +1,423 @@
...
+      description: Response shape for a pending inventory fetch request.
+      type: object
+      required:
+        - status
+        - cursor
+        - poll_after_seconds
+      properties:
+        status:
+          type: string
+          pattern: '^pending$'
```
Comment (marcusmateus) [inline packages/inventory-api/spec.yaml:201]: Interesting... is this how you are defining a "constant" in the spec? I have not seen this approach before. I have only seen the suggestion of using an enum with a single entry. Were there suggested benefits listed to this approach?

**Review by marcusmateus — APPROVED** [review]
A solid initial pass.

File: packages/inventory-api/spec.yaml (hunk 13 above)
Comment (ian-schu) [reply packages/inventory-api/spec.yaml:201]: Yep! It is the only alternative I could find to a 1-element enum. On balance, I think that other approach might be better. Not sure 🤔

**Comment by ian-schu** [comment]
@jeversmann-shipstation there's been some more churn here since you last came around, most notably my choice to handle long-running operations without a /status endpoint. I'm not sure if that's a hard-stop in your eyes, but at this point (if you'll give me a rubber stamp) what I'd like to do is merge the PR just for the sake of getting something into the repo. The idea is we're still not calling this "done", but I just don't like PRs to hang open for weeks on end.

Later today I'll be working this over with Theo and Josh F, so it will certainly be subject to more review and iteration in the days to come. I just need an approval from one of you 1st class Shipengine citizens so I can merge, if you don't mind 😏

**Review by jeversmann-shipstation — APPROVED** [review]
Handling pending transactions with the `cursor` instead of a different endpoint is an interesting trick, and I'm not opposed to it. You have my ✅ to get this in and I'll defer to Flanagan if we want to unify stuff like that between integration types.
